File: ios/HeliosApp/DesignSystem/Tokens/HeliosTypography.swift
import SwiftUI

// MARK: - HeliosTypography — SF Pro scale from design-tokens.json

enum HeliosTypography {
    enum Scale: String, CaseIterable {
        case hero
        case title1
        case title2
        case title3
        case headline
        case body
        case callout
        case subheadline
        case caption
        case caption2
    }

    struct Token {
        let size: CGFloat
        let weight: Font.Weight
        let lineHeight: CGFloat
        let tracking: CGFloat
        let font: Font.Design
    }

    static func token(for scale: Scale) -> Token {
        switch scale {
        case .hero:
            return Token(size: 56, weight: .light, lineHeight: 56, tracking: -0.02, font: .default)
        case .title1:
            return Token(size: 34, weight: .regular, lineHeight: 40, tracking: -0.01, font: .default)
        case .title2:
            return Token(size: 28, weight: .regular, lineHeight: 34, tracking: -0.01, font: .default)
        case .title3:
            return Token(size: 22, weight: .regular, lineHeight: 28, tracking: -0.01, font: .default)
        case .headline:
            return Token(size: 17, weight: .semibold, lineHeight: 22, tracking: -0.01, font: .default)
        case .body:
            return Token(size: 16, weight: .regular, lineHeight: 22, tracking: -0.01, font: .default)
        case .callout:
            return Token(size: 14, weight: .regular, lineHeight: 20, tracking: -0.01, font: .default)
        case .subheadline:
            return Token(size: 13, weight: .regular, lineHeight: 18, tracking: -0.01, font: .default)
        case .caption:
            return Token(size: 11, weight: .regular, lineHeight: 14, tracking: 0.01, font: .default)
        case .caption2:
            return Token(size: 10, weight: .medium, lineHeight: 12, tracking: 0.06, font: .default)
        }
    }

    enum NumericVariant {
        case body
        case chartAxis
        case liveTicker
    }

    static func numericFont(size: CGFloat, weight: Font.Weight = .regular) -> Font {
        let uiFont = UIFont.monospacedDigitSystemFont(ofSize: size, weight: uiWeight(weight))
        return Font(uiFont)
    }

    static func font(for scale: Scale) -> Font {
        let t = token(for: scale)
        return .system(size: t.size, weight: t.weight, design: t.font)
    }

    static func numericFont(for scale: Scale, variant: NumericVariant = .liveTicker) -> Font {
        let t = token(for: scale)
        return numericFont(size: t.size, weight: t.weight)
    }

    private static func uiWeight(_ weight: Font.Weight) -> UIFont.Weight {
        switch weight {
        case .ultraLight: return .ultraLight
        case .thin: return .thin
        case .light: return .light
        case .regular: return .regular
        case .medium: return .medium
        case .semibold: return .semibold
        case .bold: return .bold
        case .heavy: return .heavy
        case .black: return .black
        default: return .regular
        }
    }
}

// MARK: - View Modifier

struct HeliosTextStyle: ViewModifier {
    let scale: HeliosTypography.Scale
    let color: Color

    func body(content: Content) -> some View {
        let t = HeliosTypography.token(for: scale)
        return content
            .font(.system(size: t.size, weight: t.weight, design: t.font))
            .tracking(t.tracking)
            .foregroundStyle(color)
    }
}

extension View {
    func heliosTextStyle(_ scale: HeliosTypography.Scale, color: Color) -> some View {
        modifier(HeliosTextStyle(scale: scale, color: color))
    }
}

File: android/app/src/main/java/com/helios/core/designsystem/type/HeliosTypography.kt
package com.helios.core.designsystem.type

import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.em
import androidx.compose.ui.unit.sp

/**
 * Roboto Flex typography scale (10 sizes) from shared-spec/design-tokens.json.
 * Numeric variants (tabular lining) for charts and tickers.
 */
object HeliosTypography {

    val displayFontFamily = FontFamily.SansSerif // Roboto Flex fallback
    val textFontFamily = FontFamily.SansSerif
    val monoFontFamily = FontFamily.Monospace // Roboto Mono fallback

    /** Hero 56/56, weight 300, tracking -0.02em */
    val hero = TextStyle(
        fontFamily = displayFontFamily,
        fontSize = 56.sp,
        fontWeight = FontWeight.Light,
        lineHeight = 56.sp,
        letterSpacing = (-0.02).em
    )

    /** Title-1 34/40, weight 400, tracking -0.01em */
    val title1 = TextStyle(
        fontFamily = displayFontFamily,
        fontSize = 34.sp,
        fontWeight = FontWeight.Normal,
        lineHeight = 40.sp,
        letterSpacing = (-0.01).em
    )

    /** Title-2 28/34, weight 400, tracking -0.01em */
    val title2 = TextStyle(
        fontFamily = displayFontFamily,
        fontSize = 28.sp,
        fontWeight = FontWeight.Normal,
        lineHeight = 34.sp,
        letterSpacing = (-0.01).em
    )

    /** Title-3 22/28, weight 400, tracking -0.01em */
    val title3 = TextStyle(
        fontFamily = displayFontFamily,
        fontSize = 22.sp,
        fontWeight = FontWeight.Normal,
        lineHeight = 28.sp,
        letterSpacing = (-0.01).em
    )

    /** Headline 17/22, weight 600, tracking -0.01em */
    val headline = TextStyle(
        fontFamily = textFontFamily,
        fontSize = 17.sp,
        fontWeight = FontWeight.SemiBold,
        lineHeight = 22.sp,
        letterSpacing = (-0.01).em
    )

    /** Body 16/22, weight 400, tracking -0.01em */
    val body = TextStyle(
        fontFamily = textFontFamily,
        fontSize = 16.sp,
        fontWeight = FontWeight.Normal,
        lineHeight = 22.sp,
        letterSpacing = (-0.01).em
    )

    /** Callout 14/20, weight 400, tracking -0.01em */
    val callout = TextStyle(
        fontFamily = textFontFamily,
        fontSize = 14.sp,
        fontWeight = FontWeight.Normal,
        lineHeight = 20.sp,
        letterSpacing = (-0.01).em
    )

    /** Subheadline 13/18, weight 400, tracking -0.01em */
    val subheadline = TextStyle(
        fontFamily = textFontFamily,
        fontSize = 13.sp,
        fontWeight = FontWeight.Normal,
        lineHeight = 18.sp,
        letterSpacing = (-0.01).em
    )

    /** Caption 11/14, weight 400, tracking +0.01em */
    val caption = TextStyle(
        fontFamily = textFontFamily,
        fontSize = 11.sp,
        fontWeight = FontWeight.Normal,
        lineHeight = 14.sp,
        letterSpacing = 0.01.em
    )

    /** Caption-2 10/12, weight 500, tracking +0.06em */
    val caption2 = TextStyle(
        fontFamily = textFontFamily,
        fontSize = 10.sp,
        fontWeight = FontWeight.Medium,
        lineHeight = 12.sp,
        letterSpacing = 0.06.em
    )

    /** Numeric variant for charts: tabular lining figures */
    val chartNumeric = TextStyle(
        fontFamily = monoFontFamily,
        fontSize = 13.sp,
        fontWeight = FontWeight.Normal,
        lineHeight = 18.sp,
        textAlign = TextAlign.End
    )

    /** Numeric variant for live tickers: tabular lining figures */
    val liveTicker = TextStyle(
        fontFamily = monoFontFamily,
        fontSize = 28.sp,
        fontWeight = FontWeight.SemiBold,
        lineHeight = 34.sp,
        letterSpacing = (-0.01).em
    )
}
